liblouis/compileTranslationTable.c

```diff
@@ -2672,93 +2667,95 @@ compileRule(FileInfo *nested, TranslationTableHeader **table,
-		compileCharDef(nested, opcode, CTC_Space, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_Space, noback, nofor, table, displayTable);
 	case CTO_Digit:
-		compileCharDef(nested, opcode, CTC_Digit, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_Digit, noback, nofor, table, displayTable);
 	case CTO_LitDigit:
-		compileCharDef(nested, opcode, CTC_LitDigit, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_LitDigit, noback, nofor, table, displayTable);
 	case CTO_Punctuation:
-		compileCharDef(
+		return compileCharDef(
 				nested, opcode, CTC_Punctuation, noback, nofor, table, displayTable);
-		break;
 	case CTO_Math:
-		compileCharDef(nested, opcode, CTC_Math, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_Math, noback, nofor, table, displayTable);
 	case CTO_Sign:
-		compileCharDef(nested, opcode, CTC_Sign, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_Sign, noback, nofor, table, displayTable);
 	case CTO_Letter:
-		compileCharDef(nested, opcode, CTC_Letter, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_Letter, noback, nofor, table, displayTable);
 	case CTO_UpperCase:
-		compileCharDef(nested, opcode, CTC_UpperCase, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_UpperCase, noback, nofor, table, displayTable);
 	case CTO_LowerCase:
-		compileCharDef(nested, opcode, CTC_LowerCase, noback, nofor, table, displayTable);
-		break;
+		return compileCharDef(
+				nested, opcode, CTC_LowerCase, noback, nofor, table, displayTable);
 	case CTO_Grouping:
-		ok = compileGrouping(nested, noback, nofor, table, displayTable);
-		break;
+		return compileGrouping(nested, noback, nofor, table, displayTable);
 	case CTO_UpLow:
-		ok = compileUplow(nested, noback, nofor, table, displayTable);
-		break;
+		return compileUplow(nested, noback, nofor, table, displayTable);
 	case CTO_Display:
-		if (!displayTable) break;
-		if (getRuleCharsText(nested, &ruleChars))
-			if (getRuleDotsPattern(nested, &ruleDots)) {
-				if (ruleChars.length != 1 || ruleDots.length != 1) {
-					compileError(
-							nested, "Exactly one character and one cell are required.");
-					ok = 0;
+		if (displayTable) {
+			if (getRuleCharsText(nested, &ruleChars))
+				if (getRuleDotsPattern(nested, &ruleDots)) {
+					if (ruleChars.length != 1 || ruleDots.length != 1) {
+						compileError(nested,
+								"Exactly one character and one cell are required.");
+						return 0;
+					}
+					putCharDotsMapping(
+							nested, ruleChars.chars[0], ruleDots.chars[0], displayTable);
+					return 1;
 				}
-				putCharDotsMapping(
-						nested, ruleChars.chars[0], ruleDots.chars[0], displayTable);
-			}
-		break;
+			return 0;
+		}
+		return 1;
```

Comment:
Why can you have a display opcode and no display table? Shouldn't that be an error and return 0?

Reply:
The idea is to change this in the future, but at the moment it is still allowed to mix `display` rules with translation rules, and when such a table is compiled it is possible that the display part and the translation part are compiled separately, in two passes (even though usually it is done in a single pass). This is why we can't error when we're compiling a translation table and we encounter a `display` rule.